File: UnitSimulator.Core/Skills/TowerSkillSystem.cs
using System.Numerics;

namespace UnitSimulator.Skills;

public class TowerSkillSystem
{
    private readonly Dictionary<int, List<TowerSkill>> _towerSkills = new();
    
    public void RegisterSkill(int towerId, TowerSkill skill)
    {
        if (!_towerSkills.ContainsKey(towerId))
        {
            _towerSkills[towerId] = new List<TowerSkill>();
        }
        _towerSkills[towerId].Add(skill);
    }
    
    public void RegisterSkills(int towerId, IEnumerable<TowerSkill> skills)
    {
        foreach (var skill in skills)
        {
            RegisterSkill(towerId, skill);
        }
    }
    
    public TowerSkill? GetSkill(int towerId, string skillId)
    {
        if (!_towerSkills.TryGetValue(towerId, out var skills))
        {
            return null;
        }
        return skills.FirstOrDefault(s => s.Id == skillId);
    }
    
    public IReadOnlyList<TowerSkill> GetSkills(int towerId)
    {
        if (!_towerSkills.TryGetValue(towerId, out var skills))
        {
            return Array.Empty<TowerSkill>();
        }
        return skills;
    }
    
    public bool IsSkillOnCooldown(int towerId, string skillId)
    {
        var skill = GetSkill(towerId, skillId);
        return skill?.IsOnCooldown ?? false;
    }
    
    public int GetRemainingCooldown(int towerId, string skillId)
    {
        var skill = GetSkill(towerId, skillId);
        return skill?.RemainingCooldownMs ?? 0;
    }
    
    public SkillActivationResult ActivateSkill(
        int towerId,
        string skillId,
        Tower? tower,
        List<Unit> enemies,
        Vector2? targetPosition = null,
        int? targetUnitId = null)
    {
        if (tower == null)
        {
            return SkillActivationResult.CreateFailure(
                SkillErrorCodes.TowerNotFound,
                $"Tower '{towerId}' not found");
        }
        
        if (tower.IsDestroyed)
        {
            return SkillActivationResult.CreateFailure(
                SkillErrorCodes.TowerNotFound,
                $"Tower '{towerId}' is destroyed");
        }
        
        var skill = GetSkill(towerId, skillId);
        if (skill == null)
        {
            return SkillActivationResult.CreateFailure(
                SkillErrorCodes.SkillNotFound,
                $"Skill '{skillId}' not found on tower '{towerId}'");
        }
        
        if (skill.IsOnCooldown)
        {
            return SkillActivationResult.CreateFailure(
                SkillErrorCodes.SkillOnCooldown,
                $"Skill is on cooldown. Remaining: {skill.RemainingCooldownMs}ms");
        }
        
        var targetValidation = ValidateTarget(skill, targetPosition, targetUnitId, enemies);
        if (!targetValidation.IsValid)
        {
            return SkillActivationResult.CreateFailure(
                targetValidation.ErrorCode!,
                targetValidation.ErrorMessage!);
        }
        
        var effects = ApplySkillEffects(skill, tower, enemies, targetPosition, targetUnitId);
        
        skill.StartCooldown();
        
        return SkillActivationResult.CreateSuccess(skill.CooldownMs, effects);
    }
    
    public void UpdateCooldowns(int deltaMs)
    {
        foreach (var skills in _towerSkills.Values)
        {
            foreach (var skill in skills)
            {
                skill.UpdateCooldown(deltaMs);
            }
        }
    }
    
    public void UpdateCooldowns(int towerId, int deltaMs)
    {
        if (_towerSkills.TryGetValue(towerId, out var skills))
        {
            foreach (var skill in skills)
            {
                skill.UpdateCooldown(deltaMs);
            }
        }
    }
    
    private (bool IsValid, string? ErrorCode, string? ErrorMessage) ValidateTarget(
        TowerSkill skill,
        Vector2? targetPosition,
        int? targetUnitId,
        List<Unit> enemies)
    {
        switch (skill.TargetType)
        {
            case SkillTargetType.None:
                return (true, null, null);
                
            case SkillTargetType.Position:
                if (targetPosition == null)
                {
                    return (false, SkillErrorCodes.TargetRequired, 
                        "Target position is required for this skill");
                }
                return (true, null, null);
                
            case SkillTargetType.SingleUnit:
                if (targetUnitId == null)
                {
                    return (false, SkillErrorCodes.TargetRequired,
                        "Target unit is required for this skill");
                }
                var targetUnit = enemies.FirstOrDefault(u => u.Id == targetUnitId);
                if (targetUnit == null || targetUnit.IsDead)
                {
                    return (false, SkillErrorCodes.TargetNotFound,
                        $"Target unit '{targetUnitId}' not found or dead");
                }
                return (true, null, null);
                
            default:
                return (true, null, null);
        }
    }
    
    private List<SkillEffectResult> ApplySkillEffects(
        TowerSkill skill,
        Tower tower,
        List<Unit> enemies,
        Vector2? targetPosition,
        int? targetUnitId)
    {
        var effects = new List<SkillEffectResult>();
        
        switch (skill.EffectType)
        {
            case SkillEffectType.TargetedDamage:
                effects.AddRange(ApplyTargetedDamage(skill, enemies, targetUnitId));
                break;
                
            case SkillEffectType.AreaOfEffect:
                effects.AddRange(ApplyAreaDamage(skill, tower, enemies, targetPosition));
                break;
                
            case SkillEffectType.Buff:
            case SkillEffectType.Debuff:
            case SkillEffectType.Utility:
                // Phase 2: Implement buff/debuff system
                break;
        }
        
        return effects;
    }
    
    private List<SkillEffectResult> ApplyTargetedDamage(
        TowerSkill skill,
        List<Unit> enemies,
        int? targetUnitId)
    {
        var effects = new List<SkillEffectResult>();
        
        if (targetUnitId == null) return effects;
        
        var target = enemies.FirstOrDefault(u => u.Id == targetUnitId);
        if (target == null || target.IsDead) return effects;
        
        target.TakeDamage(skill.Damage);
        
        effects.Add(new SkillEffectResult
        {
            Type = "Damage",
            TargetId = target.Id.ToString(),
            Value = skill.Damage
        });
        
        return effects;
    }
    
    private List<SkillEffectResult> ApplyAreaDamage(
        TowerSkill skill,
        Tower tower,
        List<Unit> enemies,
        Vector2? targetPosition)
    {
        var effects = new List<SkillEffectResult>();
        
        var center = targetPosition ?? tower.Position;
        
        foreach (var enemy in enemies)
        {
            if (enemy.IsDead) continue;
            
            var distance = Vector2.Distance(center, enemy.Position);
            if (distance > skill.Range) continue;
            
            enemy.TakeDamage(skill.Damage);
            
            effects.Add(new SkillEffectResult
            {
                Type = "Damage",
                TargetId = enemy.Id.ToString(),
                Value = skill.Damage
            });
        }
        
        return effects;
    }
    
    public void ClearSkills(int towerId)
    {
        _towerSkills.Remove(towerId);
    }
    
    public void ClearAllSkills()
    {
        _towerSkills.Clear();
    }
}

File: UnitSimulatorUE/Plugins/UnitSimCore/Source/UnitSimCore/Private/Skills/TowerSkillSystem.cpp
#include "Skills/TowerSkillSystem.h"
#include "Units/Unit.h"
#include "Towers/Tower.h"

void FTowerSkillSystem::RegisterSkill(int32 TowerId, const FTowerSkill& Skill)
{
	TArray<FTowerSkill>& Skills = TowerSkills.FindOrAdd(TowerId);
	Skills.Add(Skill);
}

void FTowerSkillSystem::RegisterSkills(int32 TowerId, const TArray<FTowerSkill>& Skills)
{
	for (const FTowerSkill& Skill : Skills)
	{
		RegisterSkill(TowerId, Skill);
	}
}

FTowerSkill* FTowerSkillSystem::GetSkill(int32 TowerId, const FString& SkillId)
{
	TArray<FTowerSkill>* Skills = TowerSkills.Find(TowerId);
	if (Skills == nullptr) return nullptr;

	for (FTowerSkill& Skill : *Skills)
	{
		if (Skill.Id == SkillId) return &Skill;
	}
	return nullptr;
}

const TArray<FTowerSkill>* FTowerSkillSystem::GetSkills(int32 TowerId) const
{
	return TowerSkills.Find(TowerId);
}

bool FTowerSkillSystem::IsSkillOnCooldown(int32 TowerId, const FString& SkillId)
{
	const FTowerSkill* Skill = GetSkill(TowerId, SkillId);
	return Skill != nullptr && Skill->IsOnCooldown();
}

int32 FTowerSkillSystem::GetRemainingCooldown(int32 TowerId, const FString& SkillId)
{
	const FTowerSkill* Skill = GetSkill(TowerId, SkillId);
	return Skill != nullptr ? Skill->RemainingCooldownMs : 0;
}

FSkillActivationResult FTowerSkillSystem::ActivateSkill(
	int32 TowerId,
	const FString& SkillId,
	FTower* Tower,
	TArray<FUnit>& Enemies,
	const FVector2D* TargetPosition,
	int32 TargetUnitId)
{
	if (Tower == nullptr)
	{
		return FSkillActivationResult::CreateFailure(
			SkillErrorCodes::TowerNotFound,
			FString::Printf(TEXT("Tower '%d' not found"), TowerId));
	}

	if (Tower->IsDestroyed())
	{
		return FSkillActivationResult::CreateFailure(
			SkillErrorCodes::TowerNotFound,
			FString::Printf(TEXT("Tower '%d' is destroyed"), TowerId));
	}

	FTowerSkill* Skill = GetSkill(TowerId, SkillId);
	if (Skill == nullptr)
	{
		return FSkillActivationResult::CreateFailure(
			SkillErrorCodes::SkillNotFound,
			FString::Printf(TEXT("Skill '%s' not found on tower '%d'"), *SkillId, TowerId));
	}

	if (Skill->IsOnCooldown())
	{
		return FSkillActivationResult::CreateFailure(
			SkillErrorCodes::SkillOnCooldown,
			FString::Printf(TEXT("Skill is on cooldown. Remaining: %dms"), Skill->RemainingCooldownMs));
	}

	FTargetValidation Validation = ValidateTarget(*Skill, TargetPosition, TargetUnitId, Enemies);
	if (!Validation.bIsValid)
	{
		return FSkillActivationResult::CreateFailure(Validation.ErrorCode, Validation.ErrorMessage);
	}

	TArray<FSkillEffectResult> Effects = ApplySkillEffects(*Skill, *Tower, Enemies, TargetPosition, TargetUnitId);
	Skill->StartCooldown();

	return FSkillActivationResult::CreateSuccess(Skill->CooldownMs, Effects);
}

void FTowerSkillSystem::UpdateCooldowns(int32 DeltaMs)
{
	for (auto& Pair : TowerSkills)
	{
		for (FTowerSkill& Skill : Pair.Value)
		{
			Skill.UpdateCooldown(DeltaMs);
		}
	}
}

void FTowerSkillSystem::UpdateCooldowns(int32 TowerId, int32 DeltaMs)
{
	TArray<FTowerSkill>* Skills = TowerSkills.Find(TowerId);
	if (Skills == nullptr) return;

	for (FTowerSkill& Skill : *Skills)
	{
		Skill.UpdateCooldown(DeltaMs);
	}
}

void FTowerSkillSystem::ClearSkills(int32 TowerId)
{
	TowerSkills.Remove(TowerId);
}

void FTowerSkillSystem::ClearAllSkills()
{
	TowerSkills.Empty();
}

FTowerSkillSystem::FTargetValidation FTowerSkillSystem::ValidateTarget(
	const FTowerSkill& Skill,
	const FVector2D* TargetPosition,
	int32 TargetUnitId,
	const TArray<FUnit>& Enemies)
{
	switch (Skill.TargetType)
	{
	case ESkillTargetType::None:
		return { true, {}, {} };

	case ESkillTargetType::Position:
		if (TargetPosition == nullptr)
		{
			return { false, SkillErrorCodes::TargetRequired,
				TEXT("Target position is required for this skill") };
		}
		return { true, {}, {} };

	case ESkillTargetType::SingleUnit:
		if (TargetUnitId < 0)
		{
			return { false, SkillErrorCodes::TargetRequired,
				TEXT("Target unit is required for this skill") };
		}
		{
			bool bFound = false;
			for (const FUnit& Enemy : Enemies)
			{
				if (Enemy.Id == TargetUnitId && !Enemy.bIsDead)
				{
					bFound = true;
					break;
				}
			}
			if (!bFound)
			{
				return { false, SkillErrorCodes::TargetNotFound,
					FString::Printf(TEXT("Target unit '%d' not found or dead"), TargetUnitId) };
			}
		}
		return { true, {}, {} };

	default:
		return { true, {}, {} };
	}
}

TArray<FSkillEffectResult> FTowerSkillSystem::ApplySkillEffects(
	const FTowerSkill& Skill,
	const FTower& Tower,
	TArray<FUnit>& Enemies,
	const FVector2D* TargetPosition,
	int32 TargetUnitId)
{
	TArray<FSkillEffectResult> Effects;

	switch (Skill.EffectType)
	{
	case ESkillEffectType::TargetedDamage:
		Effects = ApplyTargetedDamage(Skill, Enemies, TargetUnitId);
		break;

	case ESkillEffectType::AreaOfEffect:
		Effects = ApplyAreaDamage(Skill, Tower, Enemies, TargetPosition);
		break;

	case ESkillEffectType::Buff:
	case ESkillEffectType::Debuff:
	case ESkillEffectType::Utility:
		// Phase 2: Implement buff/debuff system
		break;
	}

	return Effects;
}

TArray<FSkillEffectResult> FTowerSkillSystem::ApplyTargetedDamage(
	const FTowerSkill& Skill,
	TArray<FUnit>& Enemies,
	int32 TargetUnitId)
{
	TArray<FSkillEffectResult> Effects;
	if (TargetUnitId < 0) return Effects;

	for (FUnit& Enemy : Enemies)
	{
		if (Enemy.Id == TargetUnitId && !Enemy.bIsDead)
		{
			Enemy.TakeDamage(Skill.Damage);

			FSkillEffectResult Effect;
			Effect.Type = TEXT("Damage");
			Effect.TargetId = FString::FromInt(Enemy.Id);
			Effect.Value = Skill.Damage;
			Effects.Add(Effect);
			break;
		}
	}

	return Effects;
}

TArray<FSkillEffectResult> FTowerSkillSystem::ApplyAreaDamage(
	const FTowerSkill& Skill,
	const FTower& Tower,
	TArray<FUnit>& Enemies,
	const FVector2D* TargetPosition)
{
	TArray<FSkillEffectResult> Effects;
	const FVector2D Center = TargetPosition != nullptr ? *TargetPosition : Tower.Position;

	for (FUnit& Enemy : Enemies)
	{
		if (Enemy.bIsDead) continue;

		const float Distance = FVector2D::Distance(Center, Enemy.Position);
		if (Distance > Skill.Range) continue;

		Enemy.TakeDamage(Skill.Damage);

		FSkillEffectResult Effect;
		Effect.Type = TEXT("Damage");
		Effect.TargetId = FString::FromInt(Enemy.Id);
		Effect.Value = Skill.Damage;
		Effects.Add(Effect);
	}

	return Effects;
}
